### PR TITLE
IAO Linear Vesting

### PR DESCRIPTION
**Pull request by DeFiFoFum**
Provide a `_vestingBlockOffset` to linearly distribute 75% of the offering tokens after IAO end

IAO to IAOLinearVesting DIff: https://www.diffchecker.com/Y4hGj6Af

- [ ] UI Testing
- [ ] Audit before merge
- [x] Create IAOLinearVesting contract to vest offer tokens linearly once over
- [x] Update documentation
- [x] Provide tests

### CODE REVIEW

File: contracts/IAOLinearVesting.sol [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+        totalDebt += _amount;
+
+        emit Deposit(msg.sender, _amount);
+    }
+
+    function harvest() external nonReentrant {
+        require(block.number > endBlock, "not harvest time");
+        UserInfo storage currentUserInfo = userInfo[msg.sender];
+        require(currentUserInfo.amount > 0, "have you participated?");
+        require(currentUserInfo.lastBlockHarvested < vestingEndBlock, "nothing left to harvest");
```
Comment (ape-swap) [inline contracts/IAOLinearVesting.sol:165]: Could a user harvesting right before the `vestingEndBlock` lose one block worth of value?

Should this be `<=` ?

File: test/IAO-linear-vesting.test.js [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+  for (const account of accounts) {
+    const currentAccountSnapshot = await getUserInfoSnapshot(iaoContract, account);
+    accountSnapshots[account] = currentAccountSnapshot;
+  }
+
+  const snapshot = { contractSnapshot, accountSnapshots };
+  return snapshot;
+}
+
+describe('IAO Linear Vesting', function () {
```
Comment (ape-swap) [inline test/IAO-linear-vesting.test.js:49]: Very nice!

File: test/IAO-linear-vesting.test.js [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+
+  const accountSnapshots = {};
+  for (const account of accounts) {
+    const currentAccountSnapshot = await getUserInfoSnapshot(iaoContract, account);
+    accountSnapshots[account] = currentAccountSnapshot;
+  }
+
+  const snapshot = { contractSnapshot, accountSnapshots };
+  return snapshot;
+}
```
Comment (ape-swap) [inline test/IAO-linear-vesting.test.js:47]: Super dope

**Review by ape-swap — COMMENTED** [review]
Amazing work and tests!

I do wonder if we should do BNB testning on vested iaos too

File: contracts/IAOLinearVesting.sol (hunk 1 above)
Comment (DeFiFoFum) [reply contracts/IAOLinearVesting.sol:165]: If they harvest one block before then they would still be able to harvest on the `vestingEndBlock` as `lastBlockHarvested` would be set to the current block. Then they are still able to withdraw one more time any time in the future.

**Comment by DeFiFoFum** [comment]
> Amazing work and tests!
> 
> I do wonder if we should do BNB testning on vested iaos too

Done!

**Review by ape-swap — APPROVED** [review]
Beautiful

**Pull request merged**